Basic support for builtins from compiler-rt

vchuravy opened the pull request:
Compiler-rt is a support library for clang/llvm. It is used to support llvm intrinsics backend agnostic. Examples are soft-fp support on ARM and Float16 support.
A  motivating use-case is #17344, related issues are #14818, https://github.com/JuliaLang/julia/pull/18287#discussion_r80271796, and this will allow us to properly fix #4905.
## Caveats
- compiler-rt is often distributed alongside clang as a static library. In order to support JIT compilation we have to convert the static library into a shared object file.
- `libcompiler-rt.so` is not used by anybody, but Julia so this PR currently installs it to `usr/lib/julia`.
- Compiler-rt's build-system is coupled tightly to llvm and we would need to either build CLANG or switch to 3.8.1
- This PR to allows to build compiler-rt standalone (currently without support for atomics) following https://github.com/ReservedField/arm-compiler-rt

I am not very experienced in writing Makefiles so I would like to have some critique and I will need some help for non-Linux systems and ARM. 
## TODO
- [x] Windows support
- [x] Building and linking under MacOSX
- [x] Power support
- [x] ARM  support
- [x] ARM 32bit

vchuravy committed:
Load a shared version of compiler-rt in jitlayers

LLVM intrinsics either map to instructions or to functions in
compiler-rt. If we can't find a symbol look into a shared version
of compiler-rt and resolve the functions there.

On Darwin we have to use a unmangled version of the function name.

vchuravy committed:
Add a build-system for compiler-rt.

Compiler-rt is the runtime library of LLVM and sometimes we need to
resolve library calls to it. There are two moments were we need to
handle library calls to compiler-rt.

1. `sys.so` or any other Julia code compiled into shlibs.
  In this case we perform static linking, as with other runtime
  libraries.
2. JIT time. When we compile a Julia function that uses a libcall to crt
   we need to resolve that against a shlib copy of compiler-rt.
   Compiler-rt is usually only distributed as a static library, so we
   provide a step to turn the static library into a shared one.

The buildsystem of compiler-rt is tightly tied to LLVM. Because of this
we have a staged build-script that builds compiler-rt from source. This
standalone version should only be used for Julia.

tkelman commented on Make.inc:
```diff
@@ -228,6 +228,9 @@ LLVM_DEBUG := 0
 # set to 1 to get clang and compiler-rt
 BUILD_LLVM_CLANG := 0
+# set to 0 to not build compiler-rt standalone.
+# it will still be build if BUILD_LLVM_CLANG is set to 1
```
will still be built

tkelman commented on deps/Makefile:
```diff
@@ -39,7 +39,7 @@ include $(SRCDIR)/tools/git-external.mk
 unexport CONFIG_SITE
 
-DEP_LIBS :=
+DEP_LIBS := compiler-rt
```
only `ifeq($(BUILD_COMPILER_RT), 1)`, right?

vchuravy replied:
It is more complicated than that. As far as I can tell we will need to unconditionally build with compiler-rt (except for llvm-free builds).
`BUILD_COMPILER_RT==1` means that we will try to build a standalone compiler-rt iff LLVM hasn't build one.

tkelman commented on deps/compiler-rt.mk:
```diff
@@ -0,0 +1,144 @@
+##
+# In order to support fallbacks within llvm we need to support
+# compiler-rt. This means linking sys.so against it and resolving
+# symbols at during JIT. For the latter part we need to create
```
"symbols during JIT" ? or missing word?

tkelman commented on deps/compiler-rt.mk:
```diff
@@ -0,0 +1,144 @@
+# There are several configurations to take into account.
+# 1. USE_SYSTEM_LLVM == 1
+#    libclang_rt.builtins is distributed with clang and so
+#    we assume that USE_SYSTEM_LLVM == 1 means that clang is also
```
not sure this is a safe assumption to make

vchuravy replied:
Yes this adds an implicit dependency on clang...
We could relax this a bit by using a standalone build for `USE_SYSTEM_LLVM == 1 && BUILD_COMPILER_RT == 1`, but for `USE_SYSTEM_LLVM == 1 && BUILD_COMPILER_RT == 0`, we need to fall back onto a system version of compiler-rt and that comes at least on Archlinux with clang.

nalimilan replied:
At least for Fedora RPMs, I need  `USE_SYSTEM_LLVM=1` and I also have to build LLVM with gcc. Is that a problem, or is installing clang enough?

vchuravy replied:
Installing clang will be enough or you can set `STANDALONE_COMPILER_RT:=1`

tkelman commented on deps/compiler-rt.mk:
```diff
@@ -0,0 +1,144 @@
+#    we assume that USE_SYSTEM_LLVM == 1 means that clang is also
+#    installed.
+# 2. BUILD_LLVM_CLANG == 1
+#    libclang_rt.builtins is build with clang and we can pick that up.
```
is built with

tkelman commented on deps/compiler-rt.mk:
```diff
@@ -0,0 +1,144 @@
+# 3. BUILD_COMPILER_RT == 1
+#    Download and install compiler_rt independently of llvm/clang
+#
+# Since we need the shared opjectfile for JIT there is no USE_SYSTEM_COMPILER_RT
```
is there really no desire from upstream to fix this there?

vchuravy replied:
I can try and fix this upstream.

tkelman replied:
worth a shot, I'd rather carry patches and maybe adjust the way we build LLVM if the long-term idea is we can upstream the patches, instead of needing extra makefile wrappers here

vchuravy replied:
We will still need the standalone buildscript, but you are right it is worth a shot.

nalimilan replied:
Also, "opjectfile" -> "objectfile"

tkelman commented on deps/compiler-rt.mk:
```diff
@@ -0,0 +1,144 @@
+	$(CC) $(LDFLAGS) -shared $(fPIC) -o $@ -nostdlib $(WHOLE_ARCHIVE) -L$(CRT_DIR) -l$(CRT_STATIC_NAME) $(WHOLE_NOARCHIVE)
+else
+# The standalone compiler-rt build is based on
+# https://github.com/ReservedField/arm-compiler-rt
```
attribute completely https://github.com/ReservedField/arm-compiler-rt/blob/0fcfe9a2e4292e66fa0148479eb71f37d04381b2/LICENSE#L12-L13

vchuravy replied:
I did not reuse code, but I was inspired by the way they solved the build issue., but I can add the copyright notice

tkelman replied:
hm okay, "based on" is a bit ambiguous

tkelman commented on deps/compiler-rt.mk:
```diff
@@ -0,0 +1,144 @@
+extract-compiler-rt: $(COMPILER_RT_SRCDIR)/source-extracted
+endif
+check-compiler-rt: #NONE
+fast-check-compiler-rt: #NONE
```
`fastcheck-compiler-rt`

tkelman commented on deps/compiler-rt.mk:
```diff
@@ -0,0 +1,144 @@
+	rm -rf $(COMPILER_RT_BUILDDIR)
+compile-compiler-rt: $(COMPILER_RT_BUILDDIR)/$(COMPILER_RT_LIBFILE)
+install-compiler-rt: $(COMPILER_RT_BUILDDIR)/$(COMPILER_RT_LIBFILE)
+	cp $(COMPILER_RT_BUILDDIR)/$(COMPILER_RT_LIBFILE) $(build_private_libdir)/
```
this will need to use one of the confusing install macros to support uninstall. might also need a manifest file, however those are supposed to work

tkelman replied:
not yet addressed

vchuravy replied:
@vtjnash When I discussed this with you, you mentioned that I might not need to use the macro since I only install two files?
I am also happy to use the staged-install mechanism, but I hadn't had time to understand it yet.

tkelman commented on src/jitlayers.cpp:
```diff
@@ -308,6 +308,19 @@ void NotifyDebugger(jit_code_entry *JITCodeEntry)
+    static const char *const prefix = "__";
+    if (!compiler_rt_hdl)
+        return 0;
+    if (strncmp(name, prefix, strlen(prefix) != 0))
```
is this supposed to be `if (strncmp(name, prefix, strlen(prefix)) != 0)` ?

vchuravy commented:
Updated the PR to address the feedback given so far. I run into two major annoyances the first is that LLVM 3.9 can't build compiler-rt without clang and that the way I discover the files is problematic, because it relies on the source being already unpacked. @vtjnash suggested splitting the Makefiles, but I was hoping that somebody else might have a better solution.

tkelman commented on Make.inc:
```diff
@@ -228,8 +228,11 @@ LLVM_DEBUG := 0
-# set to 1 to build compiler-rt standalone.
-STANDALONE_COMPILER_RT := 0
+# set to 0 to disable building compiler-rt standalone.
+# Currently compiler-rt can only be build alongside clang
```
can only be built

vchuravy committed:
adds test for llvm intrinsics

One most systems (except modern mac) these tests will fail if
compiler-rt is not properly configured.

vchuravy commented:
So travis for Linux is passing. Yeah! I will see if I can get my hand on a Mac.

@yuyichao If you have time would you mind taking a look at ARM? The build-script might need some adjustments to pick up the right Arch implementations and blacklist unnecessary ones.

vchuravy commented:
So I fixed ARM&Power + Mac. The remaining platform is windows (which I don't have access to at the moment)

tkelman commented:
Try cross-compiling?

tkelman commented on Make.inc:
```diff
@@ -481,10 +486,13 @@ OBJCOPY := $(CROSS_COMPILE)objcopy
 # file extensions
 ifeq ($(OS), WINNT)
   SHLIB_EXT := dll
+  STATICLIB_EXT := lib
```
only with msvc. where are you dealing with a lib file?

vchuravy replied:
Also under mingw64? ce33683 changed that we are now linking sys.so statically against compiler-rt.

tkelman replied:
no, mingw static libraries are .a extension. compiler-rt's build system is probably full of wrong windows==msvc assumptions that should be fixed rather than worked around the wrong way here

tkelman replied:
there's a USEMSVC or some flag like that but it's a hack, you probably shouldn't worry about it. if we ever support msvc properly it'll be via cmake anyway.

vchuravy replied:
So just changing `lib` to `a` would be sufficient for now?

vchuravy commented:
So with build support for Windows, Mac, ARM, Power this is ready for another round of reviews, before I squash and rebase.

tkelman commented on Make.inc:
```diff
@@ -486,7 +486,7 @@ OBJCOPY := $(CROSS_COMPILE)objcopy
 ifeq ($(OS), WINNT)
   SHLIB_EXT := dll
-  STATICLIB_EXT := lib
+  STATICLIB_EXT := .a # MSVC is .lib
```
if comments are on the same line in a makefile variable, the value will include the trailing space - this might not break anything, but keep it in mind

vchuravy commented on src/jitlayers.cpp:
```diff
@@ -308,6 +308,27 @@ void NotifyDebugger(jit_code_entry *JITCodeEntry)
+static uint64_t resolve_compiler_rt(const char *name)
+{
+#if defined(__APPLE__)
+    static const char *const compiler_rt_lib = "@rpath/julia/libcompiler-rt";
```
@vtjnash Is this the correct form for a rpath library? Or do I need to add `.dylib`?

tkelman commented on Makefile:
```diff
@@ -374,7 +374,11 @@ endif
+ifeq ($(OS),WINNT)
+	$(INSTALL_M) $(build_private_libdir)/libcompiler-rt.$(SHLIB_EXT) $(DESTDIR)$(bindir)
+else
 	$(INSTALL_M) $(build_private_libdir)/libcompiler-rt.$(SHLIB_EXT) $(DESTDIR)$(private_libdir)
```
why isn't this just using shlibdir then?

vchuravy commented:
Ok this is nearly done. Mac test just succeeded `From worker 9:	     * llvmcall              maxrss  199.11 MB`.

I would squash this and remove the debug information (or does anybody want to keep the poetry and the history for reviewing?)

vchuravy commented:
Original version preserved at https://github.com/JuliaLang/julia/tree/vc/crt_backup

tkelman commented:
Can you move that to your fork? Would rather clean up no-longer-relevant branches from here.

This still doesn't support uninstall, does it?

vchuravy commented:
Uninstall? You mean `clean-compiler-rt`?

On Tue, 11 Oct 2016, 16:47 Tony Kelman, <notifications@github.com> wrote:

> Can you move that to your fork? Would rather clean up no-longer-relevant
> branches from here.
>
> This still doesn't support uninstall, does it?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/18734#issuecomment-252837589>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAI3agWewdyu5rVQu9H6iGM7Y99YHeAJks5qyz8BgaJpZM4KKgMs>
> .
>

tkelman commented:
no, `uninstall-compiler-rt` - remove the installed files from the install prefix. clean removes them from the build location

tkelman commented on src/jitlayers.cpp:
```diff
@@ -324,7 +324,7 @@ static uint64_t resolve_compiler_rt(const char *name)
     static const char *const mangled_prefix = "___";
-    if (strncmp(name, mangled_prefix, strlen(mangled_prefix)) != 0) {
+    if (strncmp(name, mangled_prefix, strlen(mangled_prefix)) == 0) {
       ++name;
```
indent

vchuravy commented on deps/compiler-rt.mk:
```diff
@@ -135,12 +135,14 @@ fastcheck-compiler-rt: #NONE
 configure-compiler-rt: $(COMPILER_RT_BUILDDIR)/build-configured
 clean-compiler-rt:
```
See L145.

tkelman replied:
makefiles are awful

tkelman commented:
Can I request that we hold off on merging anything LLVM build related until we're sure the buildbots are in good shape and will handle it? We haven't had nightlies in several weeks, and it looks like there are some new consistent failures within the last few days.

vchuravy commented:
As long as we do it in the next couple of weeks, sure. I have a few changes that I would like to add that are depended on this.

vchuravy commented:
I should also add that I consider this done from my side and I would value reviews and feedback.

vchuravy commented:
For a motivating use-case see https://github.com/JuliaLang/julia/tree/vc/float16_intrinsics (and the same thing would enable use to use Float128)

vchuravy commented:
My current thinking (thanks to @maleadt for the discussion) is that we should do this and #18927. This will give us access to optimised implementations on the platforms where compiler-rt is running and #18927 will give us a compatibility layer that makes sure that we can support 128bit on 32bit platforms.

tkelman commented:
Buildbots are mostly back in order, except arm and https://github.com/JuliaLang/julia/pull/18976 is needed for mac. Little annoying if the arm binaries expire (which will happen in about 10 days), but I'm going to let whoever was responsible for breaking it figure out how to fix it.

Once #18976 gets merged, rebase whichever of these PR's you'd like me to do a trial run on the buildbots and we can see if there are going to be any issues.

vchuravy commented:
Just rebased, if you could give this a try on the buildbots that would be great.

tkelman commented:
https://build.julialang.org/builders/package_osx10.9-x64/builds/912
https://build.julialang.org/builders/package_tarball32/builds/894
https://build.julialang.org/builders/package_tarball64/builds/934
https://build.julialang.org/builders/package_tarballarm/builds/859 (I expect this to fail right now)
https://build.julialang.org/builders/package_tarballppc64le/builds/136
https://build.julialang.org/builders/package_win6.2-x64/builds/858
https://build.julialang.org/builders/package_win6.2-x86/builds/780

tkelman commented:
So 32 bit Linux has issues on that buildbot. Not sure what would cause a problem there but be okay on Travis.

vchuravy commented:
Are the buildbots running on centos 5? The libc there might be older than the one on Travis. I will see if I can get a VM with Centos 5 and test it there as well.

tkelman commented:
At the moment, yes.

StefanKarpinski commented:
Bump – what's the status with this?

vchuravy commented:
I can't reproduce the failure on the buildbots locally. (in a 32bit Centos5 vm)

tkelman commented:
using gcc from where?

vchuravy commented:
https://people.centos.org/tru/devtools-2/readme
```
[vagrant@localhost ~]$ scl enable devtoolset-2 bash
[vagrant@localhost ~]$ gcc --version
gcc (GCC) 4.8.2 20140120 (Red Hat 4.8.2-15)
```

tkelman commented:
wonder if this is another case of gcc newer than 5 (6.2 on the buildbots iirc) miscompiling llvm?

vchuravy commented:
I can try building 6.1 on that VM. 
I had one problem with `4.8.2` that `ui/repl.c` needs -lstdlibc++ to compile. (And for LLVM 3.9 it needs CXXFLAGS=--std=c++11)

vchuravy commented:
I did not manage to reproduce the buildbot failure :) 

Maybe the best course of actions is to wait on #19123 and the associated buildbot changes?

tkelman commented:
Rebased this for you to pick up llvm 3.9, ran it through the buildbots. 32 bit Linux still has issues, segfaults when trying to build the docs: https://build.julialang.org/builders/package_tarball32/builds/1140/steps/make%20binary-dist/logs/stdio

And on Windows, it's trying to use pthreads. We can't switch our compiler version on Windows to one that supports pthreads for as long as we need to be able to build Windows binaries against LLVM 3.7 on the same buildbots.
```
emutls.o:emutls.c:(.text+0x13): undefined reference to `pthread_key_create'
emutls.o:emutls.c:(.text+0x8d): undefined reference to `pthread_getspecific'
emutls.o:emutls.c:(.text+0xcf): undefined reference to `pthread_once'
emutls.o:emutls.c:(.text+0xdb): undefined reference to `pthread_mutex_lock'
emutls.o:emutls.c:(.text+0xf4): undefined reference to `pthread_mutex_unlock'
emutls.o:emutls.c:(.text+0x14f): undefined reference to `pthread_setspecific'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:81: libcompiler-rt.dll] Error 1
```

vchuravy commented on deps/compiler-rt_standalone.mk:
```diff
@@ -0,0 +1,90 @@
+# Blacklist a few files we don't want to deal with
+##
+MAKEFLAGS := --no-builtin-rules
+BLACKLIST := atomic.o atomic_flag_clear.o atomic_flag_clear_explicit.o \
```
@tkelman the `emutls` failure means we need to update the blacklist to contain the related files

tkelman replied:
commit that and we could try throwing it at the buildbots again?

tkelman replied:
looks like that fixes the windows build (need to run tests locally, will see) but not whatever's wrong with the 32 bit linux buildbot

vchuravy commented:
Interesting that it seqfaults when dlopen'ing the `sys.so`. Could you check
if the imagine only contains 32bit instructions?

On Sun, 1 Jan 2017, 18:08 Elliot Saba, <notifications@github.com> wrote:

> Running on 32-bit linux (inside a Docker container) I get through
> bootstrap, but then I get a segfault when I try to run ./julia. Here's
> the backtrace:
>
> (gdb) bt
> #0  0xf7fe479d in elf_machine_rel (skip_ifunc=<optimized out>,
>     reloc_addr_arg=<optimized out>, version=<optimized out>, sym=0xf30da45c,
>     reloc=0xf313708c, map=0x805d320) at ../sysdeps/i386/dl-machine.h:312
> #1  elf_dynamic_do_Rel (skip_ifunc=<optimized out>, lazy=<optimized out>,
>     nrelative=<optimized out>, relsize=<optimized out>, reladdr=<optimized out>,
>     map=<optimized out>) at do-rel.h:137
> #2  _dl_relocate_object (scope=0x805d4d8, reloc_mode=0, consider_profiling=0)
>     at dl-reloc.c:258
> #3  0xf7fec6bf in dl_open_worker (a=0xffffce9c) at dl-open.c:435
> #4  0xf7fe8204 in _dl_catch_error (objname=objname@entry=0xffffce94,
>     errstring=errstring@entry=0xffffce98, mallocedp=mallocedp@entry=0xffffce93,
>     operate=0xf7fec370 <dl_open_worker>, args=0xffffce9c) at dl-error.c:187
> #5  0xf7febf02 in _dl_open (file=0xffffd160 "/src/julia/usr/lib/julia/sys.so",
>     mode=-2147483646, caller_dlopen=0xf7d017da <jl_dlopen+119>, nsid=<optimized out>,
>     argc=1, argv=0xffffd644, env=0xffffd64c) at dl-open.c:660
> #6  0xf7cacbf5 in dlopen_doit (a=0xffffd050) at dlopen.c:66
> #7  0xf7fe8204 in _dl_catch_error (objname=0x805d2ec, errstring=0x805d2f0,
>     mallocedp=0x805d2e8, operate=0xf7cacb80 <dlopen_doit>, args=0xffffd050)
>     at dl-error.c:187
> #8  0xf7cad30d in _dlerror_run (operate=operate@entry=0xf7cacb80 <dlopen_doit>,
>     args=args@entry=0xffffd050) at dlerror.c:163
> #9  0xf7cacc9e in __dlopen (
>     file=file@entry=0xffffd160 "/src/julia/usr/lib/julia/sys.so", mode=2) at dlopen.c:87
> #10 0xf7d017da in jl_dlopen (
>     filename=filename@entry=0xffffd160 "/src/julia/usr/lib/julia/sys.so",
>     flags=flags@entry=9) at /src/julia/src/dlload.c:81
> #11 0xf7d01a8b in jl_load_dynamic_library_ (
>     modname=modname@entry=0xffffd390 "/src/julia/usr/lib/julia/sys",
>     flags=flags@entry=9, throw_err=throw_err@entry=0) at /src/julia/src/dlload.c:179
> #12 0xf7d01b8a in jl_load_dynamic_library_e (
>     modname=0xffffd390 "/src/julia/usr/lib/julia/sys", flags=flags@entry=9)
>     at /src/julia/src/dlload.c:204
> #13 0xf7d11ddb in jl_preload_sysimg_so (fname=<optimized out>)
>     at /src/julia/src/dump.c:2367
> #14 0xf7d04558 in _julia_init (rel=<optimized out>, rel@entry=JL_IMAGE_JULIA_HOME)
>     at /src/julia/src/init.c:532
> #15 0xf7d0510c in julia_init (rel=JL_IMAGE_JULIA_HOME) at /src/julia/src/task.c:292
> #16 0x08048f7c in main (argc=<optimized out>, argv=<optimized out>)
>     at /src/julia/ui/repl.c:251
>
> This was done through the following, you can do it too if you have docker
> installed:
>
> $ docker run -ti staticfloat/julia_workerbase_ubuntu16_04:x86
> # git clone -b vc/compiler-rt https://github.com/JuliaLang/julia /src/julia
> # cd /src/julia; make -j8
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/18734#issuecomment-269896454>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAI3avgIaLIoVJN6UYQp8BB8TriJaE_1ks5rN2zygaJpZM4KKgMs>
> .
>

staticfloat commented:
Yep, looks good to me.  Using `analyze-x86` shows:

```
./analyze-x86 /src/julia/usr/lib/julia/sys.so
instructions:
 cpuid: 0        nop: 1057       call: 109406    count: 1393634
 i486:   58
 i686:   5159
 mmx:    465
 sse:    1226
 sse2:   8087
```

felixrehren commented:
Ping (this is part of equal rights for Float16, right?)

StefanKarpinski commented:
It's good to have, but it was deemed not ready for 0.6, unfortunately.

vchuravy commented:
Yes this part of my julia runtime julep, that has the goal of adding first-class support for Float16 and Float128. This got slowed down by other commitments of mine. I still have hope to have some form of this for 1.0.

StefanKarpinski commented:
@vchuravy, what's the prospect for getting this done in the next couple of months?

vchuravy commented:
Everytime I start working on this I get into shaving another Yak. I think we can get this done in the next few months, but I will first focus on Float128 and then return to this.

vchuravy commented:
superseded by #26381, while general compiler-rt support might be something we want in the long-term I rather go down the RTLIB approach.
